### PR TITLE
Correctly populate ties when loading function from string in fitbrowser

### LINKED ISSUE #31552: Fit browser not applying ties when function loaded from string
Note this functionality is quite important for the Sequential fitting in the EngDiff UI fitting tab.

### Steps to reproduce the behavior
(1) Make a workspace 
`ws = CreateSampleWorkspace(Function="User Defined", UserDefinedFunction="name=LinearBackground, A0=0.3;name=Gaussian, PeakCentre=5, Height=10, Sigma=0.7", NumBanks=1, BankPixelWidth=1, XMin=0, XMax=10, BinWidth=0.1, WorkspaceType="Event")`
(2) Plot the workspace and open the fit browser
(3) Add a peak and Tie two parameters 
- e.g. to tie centre to width (just for demonstration!) Right-click on Sigma > Tie > type f0.PeakCentre). 
- I also set Sigma to be equal to the centre beforehand though I don't know if this was necessary! You can also try this with two peaks and bind them to have the same width etc.

(4) Copy the function to a string (Setup > Manage Setup > Copy To Clipboard)
image

The string should look like this 
`name=Gaussian,Height=493.729,PeakCentre=5,Sigma=5,constraints=(444.356<Height<543.102);ties=(f0.Sigma=f0.PeakCentre)`

(5) Clear the model (Setup > Clear Model)
(6) Load the function from a string (Setup > Manage Setup >Load From String)
(7) You will see the bounds on peak height has been applied but the tie has not been applied (and the value of sigma has been reset)

This behaviour is the same with multiple peaks, non-peak functions (e.g. Linear background), itrregardless of the NumDeriv setting (which should be true if ties are added)

### PR DESCRIPTION
**Pull request by RichardWaiteSTFC**
**Description of work.**

The ties between parameters are not stored on the individual functions in a composite function as they can tie parameter across functions (unlike fixing a parameter or applying constraints). Previous code attempted to apply ties when initiating the parameter properties of the child functions, here I initiate the ties separately after the child functions browsers have been initiated. To help do this I have made public a composite function method that parses the prefix to return the function index - I thought this would be harmless and I didn't want to reproduce it.

I have removed the old code in `initParameters` that tried to apply the ties, as far as I know that branch is never entered because ties are always on the composite function - I can always replace it (it doesn't get in the way this fix).

I have put in a test in python (which seems to be where the tests for fitting live) - I check the function string loaded matches the functions string of the composite function in the browser (which is true even on master), and that the ties have been set correctly on the child property browsers (to do this I made and exposed to python a function to check the member variable m_ties isn't empty).

I think this is quite an old bug so it is not necessary to have in release 6.1

**To test:**
(1) Follow instruction in the issue
(2) Try with more complicated functions such as
`name=Gaussian,Height=100,PeakCentre=1.45125,Sigma=0.210207;name=Lorentzian,Amplitude=100,PeakCentre=3,FWHM=3;ties=(f1.Amplitude=f0.Height,f1.FWHM=f1.PeakCentre)`
(3) Remove ties, add more ties - check there is no exception
(4) Run a fit - the ties should be respected

Fixes #31552

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by robertapplin** [comment]
I have tried loading in the following function string
```
name=Gaussian,Height=487,PeakCentre=5,Sigma=5,constraints=(438.3<Height<535.7);ties=(f0.Sigma=f0.PeakCentre)
```
but the Sigma parameter is reset to a value which is not 5. Also when I perform the fit, the tie doesn't seem to be respected

**Comment by RichardWaiteSTFC** [comment]
Oh no this was working before! I do see the same thing - in fact if you just add the function and tie manually (i.e. don't load from string) it no longer respects the tie. I'll need to rebuild to be properly sure, but it might be to do with the code I removed - I'll see if putting it back fixes this.

**Comment by RichardWaiteSTFC** [comment]
I see the same problem on master (when adding the tie manually) - i.e. it adds a tie by right-clicking the browser and the fit still doesn't respect the tie. I think it's a separate issue!

**Comment by robertapplin** [comment]
What happens when you load this string?
```
name=Gaussian,Height=21.4585,PeakCentre=5,Sigma=5;name=Lorentzian,Amplitude=21.4585,PeakCentre=34,FWHM=34;ties=(f1.Amplitude=f0.Height,f1.FWHM=f1.PeakCentre,f0.Sigma=f0.PeakCentre)
```

For me the value of f0.Sigma and f1.FWHM are reset when loaded in to the function browser. But they are respected in a fit

**Comment by RichardWaiteSTFC** [comment]
> ```
> name=Gaussian,Height=21.4585,PeakCentre=5,Sigma=5;name=Lorentzian,Amplitude=21.4585,PeakCentre=34,FWHM=34;ties=(f1.Amplitude=f0.Height,f1.FWHM=f1.PeakCentre,f0.Sigma=f0.PeakCentre)
> ```

This works as expected for me all the ties are visible in the browser and all are respected  - this is what i get post-fit to a dummy data set

**Comment by RichardWaiteSTFC** [comment]
1. So when I perform a calibration which is cropped to use only the North Bank and then focus I get 3 workspaces: the North bank in TOF and both banks in dSpacing - except I don't think they are the two banks as they look exactly the same. 
2. I'm a bit confused because you can calibrate the South Bank only and then in the focussing tab crop to North Bank (presumably using the vanadium curves from the South Bank?)- surely the crop option in the focus tab depends on what calibration workspaces exist/what the previous calibration was?
3. Also can we remove a few more workspaces post-calibration - I think it's not necessary to have the `engggui_calibration_bank_` as these are summarised in a different table with 1 row/bank. And also remove the workspace for the tof(d) and residuals - actually we could avoid making workspaces at all they're only used for plotting.

**Comment by RichardWaiteSTFC** [comment]
Last commit fixes a separate bug whereby the fit algorithm was passed a function string that did not include ties if there was only one function present (i.e. it is necessary to pass the composite function string where the ties live). note that although this means the fit executes with the ties correctly set, the value of the tied parameters are not updated in the browser (but they are correct in the output workspaces). There were two @robertapplin found two other bugs which are not related to this issue:

1. Estimated FWHM overwrites the values loaded in a function string (#31738)
2. Fit browser parameters not updating post fit with a tie present on a single function (#31740)

**Review by robertapplin — DISMISSED** [review]
The original issue is fixed and there is some nice additions to the tests. I am happy for the other issues found to be fixed separately

File: qt/python/mantidqt/widgets/test/test_fitpropertybrowserbase.py [hunk 1]
```diff
@@ -80,6 +80,41 @@ def test_load_from_string_lb(self):
...
+        for prefix in property_browser.getPeakPrefixes():
+            h = property_browser.getPeakHandler(prefix)
+            # check that the ties (as opposed to fixes) have been set on the child function property handlers
+            # note that the non-fix tie string lives on the composite function but the properties whereas
+            # the tie properties (m_ties in the C++ class) are on the child's handler
+            self.assertTrue(h.hasTies())
+            # check the peak centre is fixed
+            self.assertTrue(h.ifun().isFixed(1))
+        # check constraints on last function aren't empty
+        self.assertTrue(bool(h.ifun().getConstraints()))
```
Comment (DavidFair) [inline qt/python/mantidqt/widgets/test/test_fitpropertybrowserbase.py:102]: Converting a length into a bool always makes me nervous, as it relies on the container doing the right thing

I'm going to show my inexperience with fitting but, could we instead assert the length we expect? Naively it looks like it decomposes into two constraints:
`0.18 < Sigma`
`Sigma > 0.22`

So I would expect `assertEqual(2, len(h.ifun().getConstraints())`

File: qt/widgets/common/inc/MantidQtWidgets/Common/PropertyHandler.h [hunk 2]
```diff
@@ -205,6 +205,7 @@ class EXPORT_OPT_MANTIDQT_COMMON PropertyHandler : public QObject, public Mantid
   QString updateStructureTooltip();
 
   QMap<QString, QtProperty *> getTies() { return m_ties; };
+  bool hasTies() { return !m_ties.isEmpty(); }; // added for testing purposes
```
Comment (DavidFair) [inline qt/widgets/common/inc/MantidQtWidgets/Common/PropertyHandler.h:208]: I would drop the comment, has ties is potentially useful beyond testing. 
For example developers, can check if something hasTies and act accordingly whereas this comment might steer some people away or become stale once someone uses it outside testing

File: qt/widgets/common/src/PropertyHandler.cpp [hunk 3]
```diff
@@ -187,6 +188,23 @@ class CreateAttributeProperty : public Mantid::API::IFunction::ConstAttributeVis
   QString m_name;
 };
 
+/**
+ * Populate ties on parameter properties of child functions
+ */
+void PropertyHandler::initTies() {
+  for (size_t iparam = 0; iparam < function()->nParams(); iparam++) {
```
Comment (DavidFair) [inline qt/widgets/common/src/PropertyHandler.cpp:195]: Can we hoist the `function()` call above?  C++ will execute this call each time for no benefit as it potentially has side-effects:
```c++
auto &func = function();
for (size_t iparam = 0; iparam < func->nParams(); iparam++)
```

File: qt/widgets/common/src/PropertyHandler.cpp [hunk 4]
```diff
@@ -187,6 +188,23 @@ class CreateAttributeProperty : public Mantid::API::IFunction::ConstAttributeVis
...
+    Mantid::API::ParameterTie *tie = m_fun->getTie(iparam);
+    if (tie) {
+      // get function index from prefix (second element of pair below)
+      const auto nameIndex_pair = m_cf->parseName(function()->parameterName(iparam));
```
Comment (DavidFair) [inline qt/widgets/common/src/PropertyHandler.cpp:199]: Ditto on using `func` from above

**Review by DavidFair — CHANGES_REQUESTED** [review]
Some minor nit picks, but otherwise all looks good

File: qt/widgets/common/src/PropertyHandler.cpp (hunk 3 above)
Comment (RichardWaiteSTFC) [reply qt/widgets/common/src/PropertyHandler.cpp]: Thanks, have replaced `function() `calls (and also `m_fun`) with `m_cf`

File: qt/python/mantidqt/widgets/test/test_fitpropertybrowserbase.py (hunk 1 above)
Comment (RichardWaiteSTFC) [reply qt/python/mantidqt/widgets/test/test_fitpropertybrowserbase.py]: The constraint is actually the string "0.18 < Sigma < 0.22" - I have replaced the test to assert the length of this string is correct - so no longer converting to bool.

Comment (DavidFair) [reply qt/python/mantidqt/widgets/test/test_fitpropertybrowserbase.py]: It's a good job I'm not writing these tests my attention to details... :laughing:

**Comment by DavidFair** [comment]
Lastly (and sorry I should have said whilst reviewing) do we need release notes for this? Is this something a user has ever noticed / reported
Other than that I'm happy to go with this, thanks for those changes

**Comment by RichardWaiteSTFC** [comment]
> Lastly (and sorry I should have said whilst reviewing) do we need release notes for this? Is this something a user has ever noticed / reported
> Other than that I'm happy to go with this, thanks for those changes

It has not been reported (I found it during manual testing)  - happy to add release notes though if you still want?

**Review by DavidFair — APPROVED** [review]

**Comment by DavidFair** [comment]
Nah, I'd rather get this fix in than keep it in PR purgatory
Thanks for addressing those bits :+1:

**Pull request merged**